cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -134,23 +159,53 @@ protected final String loadMessageId() {
   @Nullable
   @Override
   protected String processMessages(Iterable<ImmutablePair<String, T>> messages) throws Exception {
-    MessageTrackingIterator iterator;
+    if (this.txSize == this.getFetchSize()) {
+      return doProcess(messages);
+    }
 
-    // Process the notifications and record the message id of where the processing is up to.
-    // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
-    long timeBoundMillis = 900L * txTimeoutSeconds;
-    iterator = TransactionRunners.run(getTransactionRunner(), context -> {
-      TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
-                                                                                              timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
-      processMessages(context, trackingIterator);
-      String lastMessageId = trackingIterator.getLastMessageId();
+    TimeBoundIterator timeBoundIterator = new TimeBoundIterator<>(messages.iterator(), timeBoundMillis);
+    MessageTrackingIterator messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+    List<ImmutablePair<String, T>> currentBatch;
+    String lastMessageId = null;
+    while (timeBoundIterator.hasNext()) {
+      currentBatch = new ArrayList<>();
+      if (!messageTrackingIterator.hasNext()) {
+        // message should process in a separate txn
+        messageTrackingIterator = new MessageTrackingIterator(timeBoundIterator);
+      }
+      while (messageTrackingIterator.hasNext() && currentBatch.size() < this.txSize) {
+        currentBatch.add(messageTrackingIterator.next());
+      }
+      try {
+        lastMessageId = doProcess(currentBatch);
+        if (lastMessageId != null && !lastMessageId.equals(messageTrackingIterator.getLastMessageId())) {
```

Comment:
this is a little hard to understand, maybe would be good to add a stoppedEarly() method to the MessageTrackingIterator?

Reply:
stoppedEarly is due to timebound iterator. 
Not needed anymore with the new implementation.